app/auth/jupyterhub_auth.py

```diff
@@ -0,0 +1,115 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import jwt
+import logging
+import requests
+import re
+
+import json
+from quart import request, redirect, url_for, session
+from urllib.parse import urljoin, urlencode, parse_qs
+
+from oic import rndstr
+
+from .. import app, store
+from .web import get_key_for_user, JWT_ALGORITHM
+
+logger = logging.getLogger(__name__)
+
+
+class JupyterhubUserToken():
+
+    def process(self, request, headers):
+
+        m = re.search(r'bearer (?P<token>.+)', headers.get('Authorization', ''), re.IGNORECASE)
+        if m:
+            # logger.debug('Authorization header present, token exchange')
+            access_token = m.group('token')
+            decodentoken = jwt.decode(
+                access_token, app.config['OIDC_PUBLIC_KEY'],
+                algorithms=JWT_ALGORITHM,
+                audience=app.config['OIDC_CLIENT_ID']
+            )
+
+            jh_token = store.get(get_key_for_user(decodentoken, 'jh_access_token'))
+            headers['Authorization'] = "token {}".format(jh_token.decode())
+
+            # logger.debug('outgoing headers: {}'.format(json.dumps(headers)))
+        else:
+            # logger.debug("No authorization header, returning empty auth headers")
+            headers.pop('Authorization', None)
+
+        return headers
+
+JUPYTERHUB_OAUTH2_PATH = "/hub/api/oauth2"
+
+
+@app.route(urljoin(app.config['SERVICE_PREFIX'], 'auth/jupyterhub/login'))
+def jupyterhub_login():
+
+    state = rndstr()
+
+    session['login_seq'] += 1
+    session['jupyterhub_state'] = state
+
+    args = {
+        'client_id': app.config['JUPYTERHUB_CLIENT_ID'],
+        'response_type': 'code',
+        'redirect_uri': app.config['HOST_NAME'] + url_for('jupyterhub_get_tokens'),
```

Comment:
This needs to be under `$JUPYTERHUB_URL/$SERVICE_PREFIX`

Reply:
this is fixed by setting ```notebooks.jupyterhub.hub.services.gateway.oauth_redirect_uri=https://<HOST_NAME>/api/auth/jupyterhub/token``` in the values when deploying renku. It is already injected with the minikube-deploy script, but must be added manually for other deployments